### PR TITLE
feat: port to_from_* and load/save_binary

### PR DESCRIPTION
**Pull request by JoanFM**
Goals:

Port to_from_* converters and `load` and `save_binary` functionality.

### CODE REVIEW

File: docarray/array/array.py [hunk 1]
```diff
@@ -386,18 +448,425 @@ def traverse_flat(
...
+        :return: the binary serialization in bytes or None if _file_ctx is passed where to store
+        """
+        with io.BytesIO() as bf:
+            self._write_bytes(
+                bf=bf,
+                compress=compress,
+                protocol=protocol,
+                _show_progress=_show_progress,
+            )
+            return base64.b64encode(bf.getvalue()).decode('utf-8')
```
Comment (samsja) [inline docarray/array/array.py:627]: should not we make the 'utf-8' a parmeters of this function ?

File: docarray/array/array.py [hunk 2]
```diff
@@ -386,18 +448,425 @@ def traverse_flat(
...
+            )
+        else:
+            return cls._load_binary_all(
+                file_ctx, load_protocol, load_compress, _show_progress
+            )
+
+    def save_binary(
+        self,
+        file: Union[str, pathlib.Path],
+        protocol: str = 'pickle-array',
```
Comment (samsja) [inline docarray/array/array.py:844]: can we put the list of all the supported protocol in the docstring ? Maybe we need a enum here actually

File: docarray/array/array.py [hunk 3]
```diff
@@ -386,18 +448,425 @@ def traverse_flat(
...
+                        t, advance=1, total_size=str(filesize.decimal(_total_size))
+                    )
+
+    @classmethod
+    def load_binary(
+        cls: Type[T],
+        file: Union[str, bytes, pathlib.Path, io.BufferedReader, _LazyRequestReader],
+        protocol: str = 'pickle-array',
+        compress: Optional[str] = None,
+        _show_progress: bool = False,
```
Comment (samsja) [inline docarray/array/array.py:796]: why is it private ?

File: docarray/array/array.py [hunk 4]
```diff
@@ -386,18 +448,425 @@ def traverse_flat(
...
+                file_ctx, load_protocol, load_compress, _show_progress
+            )
+
+    def save_binary(
+        self,
+        file: Union[str, pathlib.Path],
+        protocol: str = 'pickle-array',
+        compress: Optional[str] = None,
+    ) -> None:
+        """Save DocumentArray into a binary file.
```
Comment (samsja) [inline docarray/array/array.py:847]: We need more in the docstring IMO.

Here it is the place to explain the difference between the different serilization method etc. When to pick `pickle-array` over `protobuf-array`

**Review by samsja — CHANGES_REQUESTED** [review]
couple of comments

File: docarray/array/array.py [hunk 5]
```diff
@@ -386,18 +448,425 @@ def traverse_flat(
...
+        :param _show_progress: show progress bar, only works when protocol is `pickle` or `protobuf`
+        :return: a `DocumentArray`
+        """
+        with file_ctx as fp:
+            if isinstance(fp, bytes):
+                d = fp
+            else:
+                d = fp.read()
+
+        if protocol is not None and protocol in ('pickle-array', 'protobuf-array'):
```
[the remark this answers is not in the document]
Comment (JoanFM) [reply docarray/array/array.py:682]: it is different to apply the `compression` at `array` level or `ǹon-array` level. One would not allow to do streaming

[the remark this answers is not in the document]
Comment (samsja) [reply docarray/array/array.py:682]: but all of the option have `array` at the end right ?

File: pyproject.toml [hunk 6]
```diff
@@ -20,6 +20,7 @@ types-requests = ">=2.28.11.6"
 av = {version = ">=10.0.0", optional = true}
 fastapi = {version = ">=0.87.0", optional = true }
 rich = ">=13.1.0"
+lz4 = {version= ">=1.0.0", optional = true}
 
 [tool.poetry.extras]
 common = ["protobuf"]
```
Comment (samsja) [inline pyproject.toml:26]:
```suggestion
common = ["protobuf", "lz4"]
```

**Review by samsja — APPROVED** [review]

**Pull request merged**